File: api/garmin_auth.py
"""Server-side Garmin Connect authentication handler for getbased.

This Vercel Python serverless function handles the Garmin OAuth1 → OAuth2
exchange using the proven `garth` library (v0.8.0). The JavaScript proxy
forwards `garmin_credentials` (email/password login) and
`garmin_token_refresh` (OAuth1 token refresh) payloads here so credentials
and long-lived refresh state never reach the browser.
"""

from __future__ import annotations

import base64
import json
import os
import sys
from typing import Any
from urllib.parse import urlencode, urljoin

import cloudscraper
from garth.auth_tokens import OAuth1Token, OAuth2Token
from garth.exc import MFARequired, GarminConnectAuthenticationError
from garth.http import Client
from garth.sso import exchange, login

# ── Vercel function configuration ───────────────────────────────
# Runtime: Python 3.9 (Vercel default at time of writing)
# maxDuration is configured in vercel.json as 60 seconds.

CORS_HEADERS_BASE = {
    "Access-Control-Allow-Methods": "POST, OPTIONS",
    "Access-Control-Allow-Headers": "Content-Type",
    "Cache-Control": "no-store",
    "Vary": "Origin",
}

ALLOWED_ORIGINS = {
    "https://app.getbased.health",
    "https://getbased.health",
    "https://www.getbased.health",
    "https://beta.getbased.health",
    "https://get-based.vercel.app",
    "https://get-based-managed-subscription-v2.vercel.app",
}


def _allowed_origin(origin: str) -> str:
    """Return the origin if it is allowed, otherwise ''."""
    return origin if origin in ALLOWED_ORIGINS else ""


def _cors_headers(origin: str) -> dict[str, str]:
    headers = dict(CORS_HEADERS_BASE)
    allowed = _allowed_origin(origin)
    if allowed:
        headers["Access-Control-Allow-Origin"] = allowed
    return headers


def _json_response(
    status: int,
    body: dict[str, Any],
    origin: str,
    extra_headers: dict[str, str] | None = None,
) -> dict[str, Any]:
    headers = _cors_headers(origin)
    headers["Content-Type"] = "application/json"
    if extra_headers:
        headers.update(extra_headers)
    return {
        "statusCode": status,
        "headers": headers,
        "body": json.dumps(body),
    }


def _oauth1_to_refresh_token(oauth1: OAuth1Token) -> str:
    """Serialize the OAuth1 token as a base64-encoded JSON refresh token."""
    payload = {
        "oauth_token": oauth1.oauth_token,
        "oauth_token_secret": oauth1.oauth_token_secret,
    }
    return base64.b64encode(json.dumps(payload).encode("utf-8")).decode("utf-8")


def _refresh_token_to_oauth1(refresh_token: str) -> OAuth1Token:
    """Deserialize the base64-encoded OAuth1 refresh token."""
    raw = base64.b64decode(refresh_token, validate=True)
    payload = json.loads(raw.decode("utf-8"))
    return OAuth1Token(
        oauth_token=payload["oauth_token"],
        oauth_token_secret=payload["oauth_token_secret"],
    )


def _client_with_cloudscraper() -> Client:
    """Create a new garth Client using cloudscraper to bypass Cloudflare."""
    client = Client(domain="garmin.com")
    client.sess = cloudscraper.create_scraper()
    return client


def _token_response(
    oauth1: OAuth1Token,
    oauth2: OAuth2Token,
    origin: str,
) -> dict[str, Any]:
    return _json_response(
        200,
        {
            "access_token": oauth2.access_token,
            "refresh_token": _oauth1_to_refresh_token(oauth1),
            "expires_in": oauth2.expires_in,
            "refresh_token_expires_in": getattr(
                oauth2, "refresh_token_expires_in", None
            ),
            "token_type": oauth2.token_type,
            # Garmin does not expose a stable numeric user id in the token
            # response; callers can fetch profile via connectapi if needed.
            "user_id": None,
        },
        origin,
    )


def _error_response(status: int, message: str, origin: str, extra: dict[str, Any] | None = None) -> dict[str, Any]:
    body: dict[str, Any] = {"error": message}
    if extra:
        body.update(extra)
    return _json_response(status, body, origin)


def _handle_login(body: dict[str, Any], origin: str) -> dict[str, Any]:
    email = body.get("email")
    password = body.get("password")
    if not isinstance(email, str) or not isinstance(password, str) or not email or not password:
        return _error_response(400, "Garmin login requires email and password.", origin)

    client = _client_with_cloudscraper()
    try:
        result = login(email, password, client=client, return_on_mfa=True)
    except MFARequired as exc:
        return _error_response(
            401,
            "MFA required",
            origin,
            {"mfa_required": True, "mfa_method": getattr(exc, "mfa_method", "email")},
        )
    except GarminConnectAuthenticationError as exc:
        # Avoid surfacing internal details; do not include email/password.
        return _error_response(401, "Invalid Garmin credentials.", origin)
    except Exception as exc:
        # Log the class name only — never the credentials.
        sys.stderr.write(f"Garmin login error: {type(exc).__name__}: {exc}\n")
        return _error_response(502, "Garmin authentication service unavailable.", origin)

    # garth.login() with return_on_mfa=True can return either a tuple of tokens
    # or a tuple indicating MFA is required. Normalize safely.
    if result is None:
        return _error_response(401, "Invalid Garmin credentials.", origin)

    # MFA sentinel shape used by some garth versions
    if isinstance(result, tuple) and len(result) == 2 and result[0] == "needs_mfa":
        return _error_response(
            401,
            "MFA required",
            origin,
            {"mfa_required": True, "mfa_method": "email"},
        )

    try:
        oauth1, oauth2 = result
    except Exception:
        return _error_response(502, "Unexpected Garmin authentication response.", origin)

    if not isinstance(oauth1, OAuth1Token) or not isinstance(oauth2, OAuth2Token):
        return _error_response(502, "Unexpected Garmin authentication response.", origin)

    return _token_response(oauth1, oauth2, origin)


def _handle_refresh(body: dict[str, Any], origin: str) -> dict[str, Any]:
    refresh_token = body.get("refresh_token")
    if not isinstance(refresh_token, str) or not refresh_token:
        return _error_response(400, "Garmin refresh requires refresh_token.", origin)

    try:
        oauth1 = _refresh_token_to_oauth1(refresh_token)
    except Exception:
        return _error_response(400, "Invalid Garmin refresh token.", origin)

    client = _client_with_cloudscraper()
    try:
        oauth2 = exchange(oauth1, client=client)
    except Exception as exc:
        sys.stderr.write(f"Garmin refresh error: {type(exc).__name__}: {exc}\n")
        return _error_response(502, "Garmin token refresh failed.", origin)

    if not isinstance(oauth2, OAuth2Token):
        return _error_response(502, "Unexpected Garmin refresh response.", origin)

    return _token_response(oauth1, oauth2, origin)


def _handle_proxy(body: dict[str, Any], origin: str) -> dict[str, Any]:
    """Forward a generic Garmin API request.

    The OAuth2 Bearer token is supplied by the caller in `headers`. This handler
    injects Cloudflare-evading session state if the caller did not already
    provide a usable cookie jar, and forwards the request to Garmin Connect.
    """
    url = body.get("url")
    method = body.get("method") or "GET"
    headers = body.get("headers") or {}
    req_body = body.get("body")

    if not isinstance(url, str) or not url:
        return _error_response(400, "Garmin proxy requires url.", origin)

    if not isinstance(headers, dict) or isinstance(headers, list):
        return _error_response(400, "Garmin proxy headers must be an object.", origin)

    # Only admit Garmin origins.
    if not (
        url.startswith("https://connect.garmin.com/")
        or url.startswith("https://connectapi.garmin.com/")
    ):
        return _error_response(403, "Garmin proxy URL not allowed.", origin)

    method = str(method).upper()
    if method not in {"GET", "POST", "PUT"}:
        return _error_response(405, "Garmin proxy method not allowed.", origin)

    # cloudscraper manages TLS fingerprinting and Cloudflare cookies.
    scraper = cloudscraper.create_scraper()

    # Ensure a content-type default for body-bearing methods.
    safe_headers = {str(k): str(v) for k, v in headers.items()}
    if method != "GET" and "content-type" not in {k.lower() for k in safe_headers}:
        safe_headers["Content-Type"] = "application/json"

    try:
        if method == "GET":
            upstream = scraper.get(url, headers=safe_headers, timeout=30)
        elif method == "POST":
            data = req_body if isinstance(req_body, str) else json.dumps(req_body) if req_body is not None else None
            upstream = scraper.post(url, headers=safe_headers, data=data, timeout=30)
        else:  # PUT
            data = req_body if isinstance(req_body, str) else json.dumps(req_body) if req_body is not None else None
            upstream = scraper.put(url, headers=safe_headers, data=data, timeout=30)
    except Exception as exc:
        sys.stderr.write(f"Garmin proxy error: {type(exc).__name__}: {exc}\n")
        return _error_response(502, "Garmin proxy upstream failed.", origin)

    response_headers = dict(upstream.headers)
    # Remove hop-by-hop headers that Vercel/ALB will re-add.
    for hop in ("connection", "content-length", "transfer-encoding", "upgrade", "keep-alive", "te", "trailer"):
        response_headers.pop(hop, None)
        response_headers.pop(hop.title(), None)
    response_headers.update(_cors_headers(origin))
    response_headers["Content-Type"] = response_headers.get("Content-Type", "application/json")

    return {
        "statusCode": upstream.status_code,
        "headers": response_headers,
        "body": upstream.text,
    }


def _handler(event: dict[str, Any], context: Any) -> dict[str, Any]:
    origin = ""
    try:
        headers = event.get("headers") or {}
        origin = str(headers.get("origin") or headers.get("Origin") or "").strip()
    except Exception:
        pass

    if event.get("httpMethod") == "OPTIONS":
        response_headers = _cors_headers(origin)
        if _allowed_origin(origin):
            response_headers["Access-Control-Allow-Origin"] = _allowed_origin(origin)
        return {
            "statusCode": 204,
            "headers": response_headers,
            "body": "",
        }

    if event.get("httpMethod") and event.get("httpMethod") != "POST":
        return _error_response(405, "Method not allowed. Use POST.", origin)

    try:
        raw_body = event.get("body") or "{}"
        if event.get("isBase64Encoded"):
            raw_body = base64.b64decode(raw_body).decode("utf-8")
        payload = json.loads(raw_body)
    except Exception:
        return _error_response(400, "Invalid JSON body.", origin)

    if not isinstance(payload, dict):
        return _error_response(400, "Payload must be an object.", origin)

    action = payload.get("action")
    if action == "login":
        return _handle_login(payload, origin)
    if action == "refresh":
        return _handle_refresh(payload, origin)
    if action == "proxy":
        return _handle_proxy(payload, origin)

    return _error_response(
        400,
        'Unsupported action. Use "login", "refresh", or "proxy".',
        origin,
    )


# Vercel serverless entry point for Python functions.
def handler(event: dict[str, Any], context: Any) -> dict[str, Any]:
    return _handler(event, context)

File: api/garmin_auth.js
// garmin_auth.js — Node.js serverless function for Garmin Connect authentication.
// Ports the garth library's OAuth1 → OAuth2 exchange to JS so it runs in
// Vercel's Node.js runtime (Python functions don't deploy with outputDirectory: ".").
//
// Flow: SSO login → service ticket → OAuth1 token → OAuth2 exchange.

/** @ts-check */
import crypto from 'node:crypto';

// ── Constants (from garth/sso.py) ───────────────────────────────
const CLIENT_ID = 'GCM_ANDROID_DARK';
const OAUTH_CONSUMER_URL = 'https://thegarth.s3.amazonaws.com/oauth_consumer.json';
const SSO_SUCCESSFUL = 'SUCCESSFUL';
const SSO_MFA_REQUIRED = 'MFA_REQUIRED';

const DOMAIN = 'garmin.com';

const SSO_UA =
  'Mozilla/5.0 (iPhone; CPU iPhone OS 18_7 like Mac OS X) ' +
  'AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148';

const SSO_PAGE_HEADERS = {
  'User-Agent': SSO_UA,
  Accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
  'Accept-Language': 'en-US,en;q=0.9',
  'Sec-Fetch-Mode': 'navigate',
  'Sec-Fetch-Dest': 'document',
};

const OAUTH_USER_AGENT = 'com.garmin.android.apps.connectmobile';

const ALLOWED_ORIGINS = new Set([
  'https://app.getbased.health',
  'https://getbased.health',
  'https://www.getbased.health',
  'https://beta.getbased.health',
  'https://get-based.vercel.app',
  'https://get-based-managed-subscription-v2.vercel.app',
  'https://getbased.kpmartin.com',
]);

// ── Cookie jar (manual, since Vercel Node fetch doesn't have one) ──
class CookieJar {
  constructor() {
    this.map = new Map();
  }

  /** Parse Set-Cookie header(s) and store. */
  capture(setCookieHeaders) {
    if (!setCookieHeaders) return;
    const headers = Array.isArray(setCookieHeaders) ? setCookieHeaders : [setCookieHeaders];
    for (const raw of headers) {
      if (!raw) continue;
      const parts = raw.split(';')[0].trim();
      if (!parts) continue;
      const eq = parts.indexOf('=');
      if (eq < 0) continue;
      const name = parts.slice(0, eq).trim();
      const value = parts.slice(eq + 1).trim();
      this.map.set(name, value);
    }
  }

  /** Produce a Cookie header string for outgoing requests. */
  header() {
    if (this.map.size === 0) return undefined;
    return Array.from(this.map, ([k, v]) => `${k}=${v}`).join('; ');
  }
}

// ── OAuth1 HMAC-SHA1 signing ────────────────────────────────────
// Ports oauthlib's request signing for the Garmin OAuth1 flow.

function percentEncode(str) {
  return encodeURIComponent(str)
    .replace(/[!'()*]/g, (c) => '%' + c.charCodeAt(0).toString(16).toUpperCase())
    .replace(/%20/g, '%20');
}

/**
 * Sign an OAuth1 request and return the Authorization header value.
 * Implements HMAC-SHA1 per RFC 5849 (OAuth 1.0).
 */
function signOAuth1({ method, url, params, consumerKey, consumerSecret, tokenKey, tokenSecret }) {
  // Collect all parameters (OAuth + query params)
  const allParams = { ...params };
  allParams.oauth_consumer_key = consumerKey;
  allParams.oauth_signature_method = 'HMAC-SHA1';
  allParams.oauth_timestamp = String(Math.floor(Date.now() / 1000));
  allParams.oauth_nonce = crypto.randomBytes(16).toString('hex');
  allParams.oauth_version = '1.0';
  if (tokenKey) allParams.oauth_token = tokenKey;

  // Normalize parameters: sort by key, then value
  const encoded = Object.entries(allParams)
    .map(([k, v]) => [percentEncode(String(k)), percentEncode(String(v))])
    .sort(([a1, a2], [b1, b2]) => (a1 < b1 ? -1 : a1 > b1 ? 1 : a2 < b2 ? -1 : a2 > b2 ? 1 : 0));

  const paramString = encoded.map(([k, v]) => `${k}=${v}`).join('&');

  // Build signature base string
  const baseUrl = url.split('#')[0].split('?')[0];
  const signatureBase = [
    method.toUpperCase(),
    percentEncode(baseUrl),
    percentEncode(paramString),
  ].join('&');

  // Sign with HMAC-SHA1
  const signingKey = `${percentEncode(consumerSecret)}&${percentEncode(tokenSecret || '')}`;
  const signature = crypto.createHmac('sha1', signingKey).update(signatureBase).digest('base64');

  // Build Authorization header
  const authParams = { ...allParams, oauth_signature: signature };
  return (
    'OAuth ' +
    Object.entries(authParams)
      .map(([k, v]) => `${percentEncode(k)}="${percentEncode(String(v))}"`)
      .join(',')
  );
}

// ── Fetch helper with cookie jar ────────────────────────────────
async function garminFetch(url, options, cookies, extraHeaders = {}) {
  const cookieHeader = cookies.header();
  const headers = { ...extraHeaders };
  if (cookieHeader) headers.Cookie = cookieHeader;

  const res = await fetch(url, {
    ...options,
    headers: { ...headers, ...(options?.headers || {}) },
    redirect: 'manual', // handle redirects manually to capture cookies at each step
  });

  // Capture cookies from Set-Cookie
  const setCookie = res.headers.getsetcookie?.();
  if (setCookie) cookies.capture(setCookie);
  else {
    // Node 18 fallback — getSetCookie() may not exist
    const sc = res.headers.get('set-cookie');
    if (sc) cookies.capture(sc);
  }

  return res;
}

// ── Token serialization helpers ─────────────────────────────────

function oauth1ToRefreshToken(oauth1) {
  const payload = {
    oauth_token: oauth1.oauth_token,
    oauth_token_secret: oauth1.oauth_token_secret,
  };
  return Buffer.from(JSON.stringify(payload)).toString('base64');
}

function refreshTokenToOAuth1(refreshToken) {
  const raw = Buffer.from(refreshToken, 'base64').toString('utf-8');
  return JSON.parse(raw);
}

function setExpirations(token) {
  const now = Math.floor(Date.now() / 1000);
  return {
    ...token,
    expires_at: now + (token.expires_in || 3600),
    refresh_token_expires_at: now + (token.refresh_token_expires_in || 7776000),
  };
}

// ── Main auth flows ─────────────────────────────────────────────

/**
 * Fetch OAuth consumer key/secret from S3 (garth's public consumer).
 */
let cachedConsumer = null;
async function getOAuthConsumer() {
  if (cachedConsumer) return cachedConsumer;
  const res = await fetch(OAUTH_CONSUMER_URL);
  if (!res.ok) throw new Error(`Failed to fetch OAuth consumer: ${res.status}`);
  cachedConsumer = await res.json();
  return cachedConsumer;
}

/**
 * Login to Garmin SSO and return { oauth1, oauth2 } or { mfa_required, mfa_method }.
 */
async function garminLogin(email, password) {
  const cookies = new CookieJar();
  const consumer = await getOAuthConsumer();
  const serviceUrl = `https://mobile.integration.${DOMAIN}/gcm/android`;
  const loginParams = {
    clientId: CLIENT_ID,
    locale: 'en-US',
    service: serviceUrl,
  };

  // Step 1: GET SSO sign-in page (sets cookies)
  const ssoPageUrl = new URL(`https://sso.${DOMAIN}/mobile/sso/en/sign-in`);
  ssoPageUrl.searchParams.set('clientId', CLIENT_ID);
  await garminFetch(ssoPageUrl.href, {
    method: 'GET',
  }, cookies, {
    ...SSO_PAGE_HEADERS,
    'Sec-Fetch-Site': 'none',
  });

  // Step 2: POST credentials to SSO login API
  const loginUrl = new URL(`https://sso.${DOMAIN}/mobile/api/login`);
  for (const [k, v] of Object.entries(loginParams)) loginUrl.searchParams.set(k, v);

  const loginRes = await garminFetch(loginUrl.href, {
    method: 'POST',
    headers: { ...SSO_PAGE_HEADERS, 'Content-Type': 'application/json' },
    body: JSON.stringify({
      username: email,
      password,
      rememberMe: false,
      captchaToken: '',
    }),
  }, cookies);

  const loginBody = await loginRes.json().catch(() => ({}));
  const respType = loginBody?.responseStatus?.type;

  if (respType === SSO_SUCCESSFUL) {
    const ticket = loginBody.serviceTicketId;
    return completeLogin(ticket, cookies, consumer);
  }

  if (respType === SSO_MFA_REQUIRED) {
    const mfaInfo = loginBody?.customerMfaInfo || {};
    const mfaMethod = mfaInfo.mfaLastMethodUsed || 'email';
    return { mfa_required: true, mfa_method: mfaMethod };
  }

  if (loginRes.status === 401 || respType === 'INVALID_CREDENTIALS') {
    const err = new Error('Invalid Garmin credentials.');
    err.status = 401;
    throw err;
  }

  const detail = loginBody?.responseStatus?.message || respType || 'unknown';
  const err = new Error(`SSO error: ${detail}`);
  err.status = 502;
  throw err;
}

/**
 * Complete login: get OAuth1 token via service ticket, then exchange for OAuth2.
 */
async function completeLogin(ticket, cookies, consumer) {
  // Best-effort: GET embed page to set Cloudflare LB cookie
  try {
    const embedUrl = new URL(`https://sso.${DOMAIN}/portal/sso/embed`);
    await garminFetch(embedUrl.href, {
      method: 'GET',
    }, cookies, {
      ...SSO_PAGE_HEADERS,
      'Sec-Fetch-Site': 'same-origin',
      Referer: `https://sso.${DOMAIN}/`,
    });
  } catch {
    // best-effort, ignore
  }

  // Step 3: Get OAuth1 token via service ticket
  const oauth1 = await getOAuth1Token(ticket, cookies, consumer);

  // Step 4: Exchange OAuth1 for OAuth2
  const oauth2 = await exchangeOAuth1ForOAuth2(oauth1, cookies, consumer, true);

  return { oauth1, oauth2 };
}

/**
 * Get OAuth1 token using the service ticket from SSO login.
 * Uses OAuth1-signed GET to the preauthorized endpoint.
 */
async function getOAuth1Token(ticket, cookies, consumer) {
  const baseUrl = `https://connectapi.${DOMAIN}/oauth-service/oauth/`;
  const loginUrl = `https://mobile.integration.${DOMAIN}/gcm/android`;
  const url = `${baseUrl}preauthorized?ticket=${encodeURIComponent(ticket)}&login-url=${encodeURIComponent(loginUrl)}&accepts-mfa-tokens=true`;

  // Sign the request with OAuth1
  const authHeader = signOAuth1({
    method: 'GET',
    url,
    params: {}, // query params are already in the URL
    consumerKey: consumer.consumer_key,
    consumerSecret: consumer.consumer_secret,
  });

  const res = await garminFetch(url, {
    method: 'GET',
    headers: {
      'User-Agent': OAUTH_USER_AGENT,
      Authorization: authHeader,
    },
  }, cookies);

  if (!res.ok) {
    const text = await res.text().catch(() => '');
    throw new Error(`OAuth1 token request failed: ${res.status} ${text}`);
  }

  const text = await res.text();
  // Response is URL-encoded form data (oauth_token=...&oauth_token_secret=...)
  const params = new URLSearchParams(text);
  return {
    oauth_token: params.get('oauth_token'),
    oauth_token_secret: params.get('oauth_token_secret'),
    mfa_token: params.get('mfa_token') || null,
    domain: DOMAIN,
  };
}

/**
 * Exchange OAuth1 token for OAuth2 access token.
 */
async function exchangeOAuth1ForOAuth2(oauth1, cookies, consumer, isLogin = false) {
  const baseUrl = `https://connectapi.${DOMAIN}/oauth-service/oauth/`;
  const url = `${baseUrl}exchange/user/2.0`;

  // Build form data
  const formData = new URLSearchParams();
  if (isLogin) formData.set('audience', 'GARMIN_CONNECT_MOBILE_ANDROID_DI');
  if (oauth1.mfa_token) formData.set('mfa_token', oauth1.mfa_token);

  // Sign the request with OAuth1 (including the token)
  const authHeader = signOAuth1({
    method: 'POST',
    url,
    params: {},
    consumerKey: consumer.consumer_key,
    consumerSecret: consumer.consumer_secret,
    tokenKey: oauth1.oauth_token,
    tokenSecret: oauth1.oauth_token_secret,
  });

  const res = await garminFetch(url, {
    method: 'POST',
    headers: {
      'User-Agent': OAUTH_USER_AGENT,
      'Content-Type': 'application/x-www-form-urlencoded',
      Authorization: authHeader,
    },
    body: formData.toString(),
  }, cookies);

  if (!res.ok) {
    const text = await res.text().catch(() => '');
    throw new Error(`OAuth2 exchange failed: ${res.status} ${text}`);
  }

  const token = await res.json();
  return setExpirations(token);
}

/**
 * Refresh an OAuth2 token using the stored OAuth1 refresh token.
 */
async function garminRefresh(refreshToken) {
  const oauth1 = refreshTokenToOAuth1(refreshToken);
  const consumer = await getOAuthConsumer();
  const cookies = new CookieJar();

  const oauth2 = await exchangeOAuth1ForOAuth2(oauth1, cookies, consumer, false);
  return { oauth1, oauth2 };
}

// ── Token response builder ──────────────────────────────────────

function buildTokenResponse(oauth1, oauth2) {
  return {
    access_token: oauth2.access_token,
    refresh_token: oauth1ToRefreshToken(oauth1),
    expires_in: oauth2.expires_in,
    refresh_token_expires_in: oauth2.refresh_token_expires_in || null,
    token_type: oauth2.token_type,
    user_id: null, // Garmin doesn't expose a stable user id in the token response
  };
}

// ── CORS + response helpers ─────────────────────────────────────

function corsHeaders(req) {
  const origin = req.headers.get('origin') || '';
  const allowed = ALLOWED_ORIGINS.has(origin) ? origin : '';
  const headers = {
    'Access-Control-Allow-Methods': 'POST, OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Cache-Control': 'no-store',
    'Vary': 'Origin',
  };
  if (allowed) headers['Access-Control-Allow-Origin'] = allowed;
  return headers;
}

function jsonResponse(status, body, req, extraHeaders = {}) {
  return new Response(JSON.stringify(body), {
    status,
    headers: { ...corsHeaders(req), 'Content-Type': 'application/json', ...extraHeaders },
  });
}

function errorResponse(status, message, req, extra = {}) {
  return jsonResponse(status, { error: message, ...extra }, req);
}

// ── Vercel serverless entry point ───────────────────────────────

export async function POST(req) {
  try {
    const payload = await req.json();
    if (!payload || typeof payload !== 'object') {
      return errorResponse(400, 'Payload must be an object.', req);
    }

    const action = payload.action;

    if (action === 'login') {
      const { email, password } = payload;
      if (!email || !password) {
        return errorResponse(400, 'Garmin login requires email and password.', req);
      }

      try {
        const result = await garminLogin(email, password);
        if (result.mfa_required) {
          return errorResponse(401, 'MFA required', req, {
            mfa_required: true,
            mfa_method: result.mfa_method,
          });
        }
        return jsonResponse(200, buildTokenResponse(result.oauth1, result.oauth2), req);
      } catch (err) {
        if (err.status === 401) {
          return errorResponse(401, 'Invalid Garmin credentials.', req);
        }
        console.error('Garmin login error:', err.message);
        return errorResponse(502, 'Garmin authentication service unavailable.', req);
      }
    }

    if (action === 'refresh') {
      const { refresh_token } = payload;
      if (!refresh_token) {
        return errorResponse(400, 'Garmin refresh requires refresh_token.', req);
      }

      try {
        const result = await garminRefresh(refresh_token);
        return jsonResponse(200, buildTokenResponse(result.oauth1, result.oauth2), req);
      } catch (err) {
        console.error('Garmin refresh error:', err.message);
        return errorResponse(502, 'Garmin token refresh failed.', req);
      }
    }

    return errorResponse(400, 'Unsupported action. Use "login" or "refresh".', req);
  } catch (err) {
    console.error('Garmin auth handler error:', err.message);
    return errorResponse(500, 'Internal server error.', req);
  }
}

export async function OPTIONS(req) {
  return new Response(null, {
    status: 204,
    headers: corsHeaders(req),
  });
}

export const config = {
  maxDuration: 60,
};